A. Build a baseline model

- One hidden layer of 10 nodes, and a ReLU activation function

- Use the adam optimizer and the mean squared error  as the loss function.

In [1]:
#First we import all necessary libraries for this project.
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
import keras
from keras.models import Sequential
from keras.layers import Dense

In [3]:
#Import our dataset
df = pd.read_csv(r"C:\Users\kdhou\Downloads\concrete_data.csv")
df.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [9]:
#Split data into features and target variables
features = df[df.columns[:-1]]
target = df[df.columns[-1]]
n_features = features.shape[1]
n_features

8

In [14]:
#Define the baseline model as a function
def baseline_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_features,)))
    model.add(Dense(1))
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    return(model)

1. Randomly split the data into a training and test sets by holding 30% of the data for testing. You can use the 
train_test_split helper function from Scikit-learn.

In [15]:
#Split data into train-test sets
x_train, x_test, y_train, y_test = train_test_split(features, target, test_size = 0.3, random_state = None)
print(x_train.shape)
print(y_train.shape)
print(x_test.shape)
print(y_test.shape)

(721, 8)
(721,)
(309, 8)
(309,)


2. Train the model on the training data using 50 epochs.

In [16]:
#Call the model
model = baseline_model()

#Fit the model as per above
model.fit(x_train, y_train, epochs = 50, verbose = 1)

Epoch 1/50


c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 642us/step - loss: 19377.2422
Epoch 2/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - loss: 2770.7161
Epoch 3/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 512us/step - loss: 1317.8043
Epoch 4/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 509us/step - loss: 1258.6567
Epoch 5/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 460us/step - loss: 1196.0315
Epoch 6/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 510us/step - loss: 1122.0908
Epoch 7/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 507us/step - loss: 1086.1405
Epoch 8/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 458us/step - loss: 1096.1532
Epoch 9/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 505us/step - loss: 1010.5873
Epoch 10/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 505us/step - loss: 998.7238
Epoch 11/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 506us/step - loss: 965.0542
Epoch 12/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 464us/step - loss: 903.8542
Epoch 13/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 466us/step - loss: 886.1525
Epoch 14/50
23/23 ━━━━━━━━━━━━━━━━━━━━ 0s 462us/step - loss: 832.0508
Epoch 15/50
23/23 ━━━━━━━━━━━━

3. Evaluate the model on the test data and compute the mean squared error between the predicted concrete strength and the actual concrete strength. You can use the mean_squared_error function from Scikit-learn.

In [23]:
#Call the evaluate method on the testing data
score = model.evaluate(x_test, y_test, verbose = 0)
print('Mean squared error of the model:', score)

Mean squared error of the model: 224.60447692871094


4. Repeat steps 1 - 3, 50 times, i.e., create a list of 50 mean squared errors.

5. Report the mean and the standard deviation of the mean squared errors.

In [25]:
#Create initial empty list
mean_squared_errors = []

#Loop through 50 iterations and append results to list
i = 1
while i <= 50:
    #Create new train/test split each iteration
    x_train, x_test, y_train, y_test = train_test_split(features, target, test_size = 0.3, random_state = None)

    #Call model function on new splits
    model = baseline_model()
    model.fit(x_train, y_train, epochs = 50, verbose = 0)

    #Evaluate on test data
    mse = model.evaluate(x_test, y_test, verbose = 0)

    #Append to list and increment i
    mean_squared_errors.append(mse)
    i += 1

#Calculate mean and stdev of MSE's
mean = np.average(mean_squared_errors)
stdev = np.std(mean_squared_errors)
print("Mean of the MSE's is: {} \n Standard Deviation of the MSE's is: {}".format(mean, stdev))

c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in

Mean of the MSE's is: 366.06535598754886 
 Standard Deviation of the MSE's is: 432.01970797020306


We see that the average MSE of the output is approximately 366.1, and the standard deviation of these MSE's is approximately 432.0. 

Let's see if we can reduce this error.

B. Normalize the data (5 marks)

Repeat Part A but use a normalized version of the data. Recall that one way to normalize the data is by subtracting the mean from the individual predictors and dividing by the standard deviation.

How does the mean of the mean squared errors compare to that from Step A?

In [29]:
#Normalize by subtracting mean and dividing by standard deviation
#Note that we could normalize the target variable, but it is not necessary.
features_normalized = (features - features.mean())/features.std()
features_normalized.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [30]:
#Pass the normalized data through the model and find the mean and stdev of the MSE's as before
#Create initial empty list
mean_squared_errors = []

#Loop through 50 iterations and append results to list
i = 1
while i <= 50:
    #Create new train/test split each iteration, this time using normalized features data
    x_train, x_test, y_train, y_test = train_test_split(features_normalized, target, test_size = 0.3, random_state = None)

    #Call model function on new splits
    model = baseline_model()
    model.fit(x_train, y_train, epochs = 50, verbose = 0)

    #Evaluate on test data
    mse = model.evaluate(x_test, y_test, verbose = 0)

    #Append to list and increment i
    mean_squared_errors.append(mse)
    i += 1

#Calculate mean and stdev of MSE's
mean = np.average(mean_squared_errors)
stdev = np.std(mean_squared_errors)
print("Mean of the MSE's is: {} \n Standard Deviation of the MSE's is: {}".format(mean, stdev))

c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in

Mean of the MSE's is: 340.0970825195312 
 Standard Deviation of the MSE's is: 98.63189180185826


Our average MSE for normalized data is approximately 340.1. Compared to an average of 366.1 with the unnormalized data, this is better but not by much.

However, our standard deviation went from 432.0 down to only 98.6, a considerable decrease. So we now have much less variation in our error term, which is good.

C. Increate the number of epochs (5 marks)

Repeat Part B but use 100 epochs this time for training.

How does the mean of the mean squared errors compare to that from Step B?

In [31]:
#We simply repeat the same code on the normalized data as in part B, but using epochs = 100 on the model.fit() method.
#Create initial empty list
mean_squared_errors = []

#Loop through 50 iterations and append results to list
i = 1
while i <= 50:
    #Create new train/test split each iteration, this time using normalized features data
    x_train, x_test, y_train, y_test = train_test_split(features_normalized, target, test_size = 0.3, random_state = None)

    #Call model function on new splits
    model = baseline_model()
    model.fit(x_train, y_train, epochs = 100, verbose = 0)

    #Evaluate on test data
    mse = model.evaluate(x_test, y_test, verbose = 0)

    #Append to list and increment i
    mean_squared_errors.append(mse)
    i += 1

#Calculate mean and stdev of MSE's
mean = np.average(mean_squared_errors)
stdev = np.std(mean_squared_errors)
print("Mean of the MSE's is: {} \nStandard Deviation of the MSE's is: {}".format(mean, stdev))

c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in

Mean of the MSE's is: 165.3788101196289 
Standard Deviation of the MSE's is: 20.264702654633


Our Mean of the MSE's has now dropped from 340.1 down to 165.4, which is huge. Our standard deviation also dropped from 98.6 down to 20.3.

We can conclude that increasing the iterations (epochs) to 100 has improved the model significantly.

Repeat part B but use a neural network with the following instead:

- Three hidden layers, each of 10 nodes and ReLU activation function.

How does the mean of the mean squared errors compare to that from Step B?

In [32]:
#We first define the new model as three_layer_model
def three_layer_model():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(n_features,)))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(10, activation='relu'))
    model.add(Dense(1))
    model.compile(optimizer = 'adam', loss = 'mean_squared_error')
    return(model)

In [33]:
#Apply this model using the code from step C to get our final output
#Create initial empty list
mean_squared_errors = []

#Loop through 50 iterations and append results to list
i = 1
while i <= 50:
    #Create new train/test split each iteration, this time using normalized features data
    x_train, x_test, y_train, y_test = train_test_split(features_normalized, target, test_size = 0.3, random_state = None)

    #Call model function on new splits
    model = three_layer_model()
    model.fit(x_train, y_train, epochs = 100, verbose = 0)

    #Evaluate on test data
    mse = model.evaluate(x_test, y_test, verbose = 0)

    #Append to list and increment i
    mean_squared_errors.append(mse)
    i += 1

#Calculate mean and stdev of MSE's
mean = np.average(mean_squared_errors)
stdev = np.std(mean_squared_errors)
print("Mean of the MSE's is: {} \nStandard Deviation of the MSE's is: {}".format(mean, stdev))

c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
c:\Users\kdhou\AppData\Local\Programs\Python\Python312\Lib\site-packages\keras\src\layers\core\dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in

Mean of the MSE's is: 89.16071472167968 
Standard Deviation of the MSE's is: 21.73842032304967


Final Mean of MSE's is 89.16, which is by far the lowest error we have seen. Standard deviation is 21.7 which is comparable to our output in part C.

Overall we can conclude that adding hidden layers has greatly improved the model by reducing the error term.